src/python/pants/engine/fs.py

```diff
@@ -125,6 +125,11 @@ class DirectoryToMaterialize(datatype([('path', text_type), ('directory_digest',
   """A request to materialize the contents of a directory digest at the provided path."""
   pass
 
+
+class UrlToFetch(datatype([('url', text_type)])):
```

Comment:
I would love to require a digest here (and to then expose it in https://github.com/pantsbuild/pants/pull/6660/files#diff-11db197014e1e80c9551e9b61aba6aa0R1018).

But since it requires a bit of design work, I've opened https://github.com/pantsbuild/pants/issues/6672, and we can leave a TODO here.

Reply:
After reviewing #6661 I realized that it would be feasible to require the `Digest` now, because the initial usecase for file downloading is downloading a universal/cross-platform pex. That sidesteps the whole "needing to configure digests for a bunch of different binaries" issue.

It _might_ be a good idea to have and use the Digest in this first PR, because I think that it would clean up where/how the file is stored and cached. If we just totally skip storing/caching in the first PR, then maybe it would be better to add the required digest as a small followup PR... smaller than #6672 even, because #6661 doesn't need to be cross-platform.

Having typed this out, tiny followup PR seems very reasonable, but interested in your thoughts.

Reply:
I have no strong feelings either way, but I feel like it would be better to have it in this PR, thinking about when people look back, I think it makes sense to put it in the same commit if we have decided that this is the behavior we want and the change is so tiny. If I were to look at the history, I think I would expect it to be in this PR.